### PR TITLE
Upload translation of Asynchronous Programming (Chapter 11)

### PR DESCRIPTION
**Pull request by Zheng7426**
@yukie1412 Hi Yukie, please kindly review the translation when you have time. Thanks. Apologies for taking so long.

### CODE REVIEW

File: 11_async.md [hunk 1]
```diff
@@ -1,190 +1,101 @@
...
-executed depends pretty much entirely on the speed of the processor.
+一个计算机的核心部分，也就是执行组成我们的程序的每个步骤的部分，被称为 _((处理器))_。我们目前所接触到的程序在它们完成工作之前都会让处理器保持繁忙。对于操纵数字的循环这样的程序，它的执行速度几乎完全取决于处理器的速度。
 
 {{index [memory, speed], [network, speed]}}
 
-But many programs interact with things outside of the processor. For
-example, they may communicate over a computer network or request
-data from the ((hard disk))—which is a lot slower than getting it from
-memory.
+但是，很多程序于处理器之外的事物进行交互。比如说，它们可能通过一个计算机网络进行沟通，或者向一个((硬盘))请求数据 —— 这比从内存获取数据要慢得多。
```
Comment (ghost) [inline 11_async.md]: should it be 与 instead?

File: 11_async.md [hunk 2]
```diff
@@ -1,190 +1,101 @@
...
 
-When such a thing is happening, it would be a shame to let the
-processor sit idle—there might be some other work it could do in the
-meantime. In part, this is handled by your operating system, which
-will switch the processor between multiple running programs. But that
-doesn't help when we want a _single_ program to be able to make
-progress while it is waiting for a network request.
+当这样的事情发生之际，让处理器处于闲置的状态是很可惜的 —— 可能它同时还有别的可以做的事情。在某种程度上，这是你的操作系统要处理的事情，它会在多个运行着的程序之间切换处理器。但是当我们只有 _单个_ 程序在等待网络请求并需要有进展的时候，这就帮不到忙了。
```
Comment (ghost) [inline 11_async.md]: it sounds weird to me. what's your opinion to rephrase it to something like （与此同时）它或许可以处理其他事情 or 它可以做些其他工作？

File: 11_async.md [hunk 3]
```diff
@@ -1,190 +1,101 @@
...
-spends running normally, and the thin lines represent time spent
-waiting for the network. In the synchronous model, the time taken by
-the network is _part_ of the timeline for a given thread of control.
-In the asynchronous model, starting a network action conceptually
-causes a _split_ in the timeline. The program that initiated the
-action continues running, and the action happens alongside it,
-notifying the program when it is finished.
+在下图中，粗线表示程序在正常运行时所花的时间，而细线表示等待网络所花费的时间。在同步模型中，对于一个给定的控制线程来说，网络所花费的时间是时间线的 _一部分_。在异步模型中，启动一个网络操作在概念上会导致时间线出现一个 _分岔_。启动该操作的程序会继续运行，而发生的操作会与其并行，且在操作完成时通知该程序。
 
 {{figure {url: "img/control-io.svg", alt: "Control flow for synchronous and asynchronous programming",width: "8cm"}}}
```
Comment (ghost) [inline 11_async.md]: let's translate this as well, since it describes the image.

File: 11_async.md [hunk 4]
```diff
@@ -1,190 +1,101 @@
...
-nests that have an unbroken visual connection can communicate.
-
-Our friend the corvid expert has mapped the network of crow nests in
-the village of ((Hières-sur-Amby)), on the banks of the river Rhône.
-This map shows the nests and their connections:
+另一种阐明两者区别的方式在于，在同步模型中，等待操作完成是 _隐性的_。而在异步模型中，它是 _显性的_、可以被我们控制的。
+
+异步性是一把双刃剑，它使得表达不符合直线控制模型的程序更加简洁，然而它也会让符合直线模型的程序表达起来更加尴尬。我们会在本章稍后的内容中介绍一些应对这种尴尬处境的办法。
+
+两个重要的 JavaScript 编程平台 —— ((浏览器)) 和 ((Node.js)) —— 都能将需要花上一段时间的操作异步化，而不是依赖于((线程))。由于线程编程是出了名的困难(要弄懂一个同时在做很多件事的程序是难上加难)，这通常被认为是一件好事。
```
Comment (ghost) [inline 11_async.md]: personally think the sentence flows better without 能，thoughts?

File: 11_async.md [hunk 5]
```diff
@@ -1,190 +1,101 @@
...
+
+然而大多数人不知道的是，它们其实还可以做很多事情，只是将其很好地隐藏了起来。一个有名望(或者说有点古怪)的((乌鸦))专家曾经告诉过我，乌鸦的技术并非远远落后于人类，而且它们正在迎头赶上我们。
+
+比方说，很多乌鸦文化拥有构建计算设备的能力。这些不像人类的电子计算设备，而是通过一种微型昆虫的动作进行操作，这类昆虫在物种上与((白蚁))很接近，它们与乌鸦发展出了((共生关系))。乌鸦给它们提供食物，作为回报，昆虫们则建立起它们复杂的殖民地，并借助居住其中的活物的帮助来执行计算。
+
+这样的殖民地通常位于庞大的、长存的鸟巢中。乌鸦和昆虫同心协力建起了一个以鳞茎状的粘土为结构的网络，隐匿于鸟巢的细枝之间，昆虫就在这地方生活和工作。
+
+为了与其他设备进行通信，这些机器使用光信号。乌鸦把部分反光材料嵌入了专用的通信杆中，而昆虫则瞄准这些通信杆来把光反射至另一个鸟巢，并以快速闪烁的序列为数据编码。这意味着只有在视线没有被挡住的鸟巢之间才可以进行通信。
+
+我们的朋友乌鸦专家已经将 ((Hières-sur-Amby)) 这个村子里的乌鸦巢的网络给映射了出来，就在 Rhône 河的岸边。这个映射展现了这些鸟巢和它们彼此之间的连结：
```
Comment (ghost) [inline 11_async.md]: 罗纳河

File: 11_async.md [hunk 6]
```diff
@@ -1,190 +1,101 @@
...
 
 {{figure {url: "img/Hieres-sur-Amby.png", alt: "A network of crow nests in a small village"}}}
```
Comment (ghost) [inline 11_async.md]: should translate

File: 11_async.md [hunk 7]
```diff
@@ -249,91 +140,47 @@ defineRequestType("note", (nest, content, source, done) => {
...
-
-Working with abstract concepts is often easier when those concepts can
-be represented by ((value))s. In the case of asynchronous actions, you
-could, instead of arranging for a function to be called at some point
-in the future, return an object that represents this future event.
+传递给处理器的第四个参数 `done` 是一个回调函数，处理器在处理完请求之后必须调用该函数。如果我们将处理器的((返回值))作为响应的值，那意味着一个请求处理器本身无法执行异步操作。一个做异步操作的函数通常在工作结束之前就返回，且已安排了一个在结束时就调用的回调函数。因此，我们需要一些异步的机制 —— 在这个情况下，再来一个((回调函数)) —— 以在响应就绪的时候发出信号。
+
+从某种程度上来说，异步性是 _可传染的_。对于一个以异步方式工作的函数，任何一个调用它的函数自身也必须是异步的，且使用一个回调或类似的机制来呈现其结果。与只是简单地返回一个值相比，调用一个回调要更为复杂，也更容易出错。因此，需要以这种方式构建你大部分的程序并不是一个好方法。
+
+## Promise
```
Comment (yukie1412) [inline 11_async.md:151]: // todo

File: 11_async.md [hunk 8]
```diff
@@ -199,44 +110,24 @@ bigOak.readStorage("food caches", caches => {
...
-callback-based functions for communication. Nests have a `send` method
-that sends off a request. It expects the name of the target nest, the
-type of the request, and the content of the request as its first three
-arguments, and it expects a function to call when a response comes in as its
-fourth and last argument.
+`"./crow-tech"` 模块导出的接口提供了基于回调的函数来进行通信。鸟巢拥有一个可以发送请求的 `send` 方法。该方法预期一个目标鸟巢的名称、请求的类型，以及请求的内容作为它的前三个参数，还预期一个在收到响应时就调用的函数作为其第四个(也是最后一个)参数。
 
 ```
 bigOak.send("Cow Pasture", "note", "Let's caw loudly at 7PM",
             () => console.log("Note delivered."));
```
Comment (ghost) [inline 11_async.md]: discussion: should we translate the strings here?

File: 11_async.md (hunk 2 above)
Comment (Zheng7426) [reply 11_async.md]: Sure, let us rephrase it to（与此同时）它或许可以处理其他事情

File: 11_async.md (hunk 3 above)
Comment (Zheng7426) [reply 11_async.md]: Good idea.

File: 11_async.md (hunk 4 above)
Comment (Zheng7426) [reply 11_async.md]: Sure, I will delete 能

File: 11_async.md (hunk 5 above)
Comment (Zheng7426) [reply 11_async.md]: 👌

File: 11_async.md (hunk 8 above)
Comment (Zheng7426) [reply 11_async.md]: If we translate the code block, would it affect other language versions when it got merged? If other versions would be intact, then I reckon we could it.

File: 11_async.md (hunk 1 above)
Comment (Zheng7426) [reply 11_async.md]: yes

File: 11_async.md (hunk 8 above)
Comment (ghost) [reply 11_async.md]: why would it affect other languages? my understanding is this branch is specific to chinese

File: 11_async.md (hunk 7 above)
Comment (ghost) [inline 11_async.md]: should this title be translated? (i'm not sure if "promise" has a chinese name)
if not translated, the original title is `## Promises`

File: 11_async.md [hunk 9]
```diff
@@ -249,91 +140,47 @@ defineRequestType("note", (nest, content, source, done) => {
...
-The easiest way to create a promise is by calling `Promise.resolve`.
-This function ensures that the value you give it is wrapped in a
-promise. If it's already a promise, it is simply returned—otherwise,
-you get a new promise that immediately finishes with your
-value as its result.
+想要创建一个 promise，最简单的办法是调用 `Promise.resolve`。这个函数会确保你给它的值被包装在一个 promise 中。如果该值已经是一个 promise，那么它会被直接返回 —— 否则，你将会得到一个立即完成的新 promise，它的结果即是你的值。
 
 ```
 let fifteen = Promise.resolve(15);
 fifteen.then(value => console.log(`Got ${value}`));
```
Comment (ghost) [inline 11_async.md]: if we localize "Got", then the comment below (line 166) can be translated directly

File: 11_async.md [hunk 10]
```diff
@@ -346,86 +193,39 @@ storage(bigOak, "enemies")
...
-This asynchronous function returns a meaningful value. This is the
-main advantage of promises—they simplify the use of asynchronous
-functions. Instead of having to pass around callbacks, promise-based
-functions look similar to regular ones: they take input as arguments
-and return their output. The only difference is that the output may
-not be available yet.
+这个异步函数会返回一个有意义的值。这是 promise 的一个主要优势 —— 它们简化了异步函数的使用方法。与其必须通过回调来传递，基于承诺的函数与正常函数看起来很像：它们接受输入作为参数，并且返回其输出。唯一的区别在于其输出在当下或许还不可用。
 
-## Failure
+## 故障
```
Comment (yukie1412) [inline 11_async.md:198]: // todo

File: 11_async.md (hunk 8 above)
Comment (Zheng7426) [reply 11_async.md]: Ok, I will translate the text inside the code block then.

File: 11_async.md (hunk 7 above)
Comment (Zheng7426) [reply 11_async.md]: Good question. Originally I translated it into 承诺. So some of the paragraphs looked like this: 

你可以使用承诺的 `then` 方法来获取其结果。在承诺被兑现并生成一个值的时候，由该方法注册的((回调函数))将会被调用。你可以向单个承诺添加多个回调，即使你是在该承诺 _已兑现_ (已完成)之后才添加的回调，它们也仍会被调用。

但那并非 `then` 方法的全部所作所为，它还返回另外一个承诺。这个承诺会兑现成处理器函数所返回的值，如果处理器函数返回的是一个承诺，那么它将等待这个承诺并将其结果给兑现。

我们不妨将承诺看成是一个可以将值带入异步现实的设备。一个普通的值是存在的，一个承诺值则 _可能_ 已经存在，或者可能会在未来的某一时间点才出现。基于承诺而定义的计算会在此类被包装的值上运作，并且在其值可用的时候被异步执行。

However, I checked the MDN documentation and found that they did not translate the word 'promise'. So my hypothesis is that in convention there is no official translation for this word, therefore I changed 承诺 back to 'promise'. What's your opinion?

Comment (ghost) [reply 11_async.md]: let's keep it as "promise" then, to be in sync w/ MDN

Comment (Zheng7426) [reply 11_async.md]: Ok👌

File: 11_async.md (hunk 9 above)
Comment (Zheng7426) [reply 11_async.md]: sounds good

File: 11_async.md [hunk 11]
```diff
@@ -163,7 +163,7 @@ defineRequestType("note", (nest, content, source, done) => {
 ```
 let fifteen = Promise.resolve(15);
 fifteen.then(value => console.log(`Got ${value}`));
```
Comment (ghost) [inline 11_async.md]: you forgot to translate line 165

File: 12_language.md [hunk 12]
```diff
@@ -1,66 +1,44 @@
 {{meta {load_files: ["code/chapter/12_language.js"], zip: "node/html"}}}
 
-# Project: A Programming Language
```
Comment (ghost) [inline 12_language.md:3]: can you have chap 12 in another PR?

File: 11_async.md [hunk 13]
```diff
@@ -1039,72 +680,44 @@ class Group {
...
 
-No matter how closely together events—such as timeouts or incoming
-requests—happen, a JavaScript environment will run only one program at
-a time. You can think of this as it running a big loop _around_ your
-program, called the _event loop_. When there's nothing to be done,
-that loop is stopped. But as events come in, they are added to a queue,
-and their code is executed one after the other. Because no two things
-run at the same time, slow-running code might delay the handling of
-other events.
+无论事件之间 —— 诸如 timeouts(译者注：超时) 或者传入的请求 —— 有多近，一个 JavaScript 环境一次只会运行一个程序。你可以将其视为一个运行着的、_围绕_ 着你的程序的大循环，名为 _事件循环_。当没有事情需要被完成的时候，这个循环就会停止。但是当事件传入时，它们会被添加到一个队列中，而且它们的代码会一个接一个地被执行。由于不会有两件事同时运行，缓慢运行的代码可能会使其他事件的处理延后。
```
Comment (ghost) [inline 11_async.md]: i think we can just use 超时 here, because `timeouts` isn's js specific.

File: 11_async.md [hunk 14]
```diff
@@ -1113,39 +726,30 @@ setTimeout(() => {
 }, 20);
 while (Date.now() < start + 50) {}
 console.log("Wasted time until", Date.now() - start);
```
Comment (ghost) [inline 11_async.md]: again, if we translate these strings here, then in comments below, we can just show the translation text

File: 11_async.md [hunk 15]
```diff
@@ -1227,72 +814,47 @@ async function chicks(nest, year) {
...
 }
 
 function locateScalpel2(nest) {
-  // Your code here.
+  // 在这里输入你的代码
 }
 
 locateScalpel(bigOak).then(console.log);
-// → Butcher Shop
+// → Butcher Shop     屠宰门店
```
Comment (ghost) [inline 11_async.md]: i don't think we need the english

File: 11_async.md [hunk 16]
```diff
@@ -1301,66 +863,42 @@ if}}
...
   return new Promise((resolve, reject) => {
-    // Your code here.
+    // 在这里输入你的代码
   });
 }
 
-// Test code.
+// 测试代码
 Promise_all([]).then(array => {
   console.log("This should be []:", array);
```
Comment (ghost) [inline 11_async.md]: let's translate the strings in these functions as well please

File: 12_language.md (hunk 12 above)
Comment (Zheng7426) [reply 12_language.md:3]: Sure. I accidentally pushed it, my bad.

File: 11_async.md (hunk 13 above)
Comment (Zheng7426) [reply 11_async.md]: okay

File: 12_language.md (hunk 12 above)
Comment (Zheng7426) [reply 12_language.md:3]: I think I kinda messed it up. I deleted chapter 12(probably should not have done so), I do have a copy of translated text for chapter 12. Once you merge chapter 11, I would create a pull request and upload chapter 12.

Comment (Zheng7426) [reply 12_language.md:3]: I think I kinda messed it up. I deleted chapter 12(probably should not have done so), I do have a copy of translated text for chapter 12. Once you merge chapter 11, I would create a pull request and upload chapter 12.

**Comment by Zheng7426** [comment]
I think I kinda messed it up. I deleted chapter 12(probably should not have done so), I do have a copy of translated text for chapter 12. Once you merge chapter 11, I would create a pull request and upload chapter 12.

File: 11_async.md [hunk 17]
```diff
@@ -715,19 +715,19 @@ try {
...
 setTimeout(() => {
-  console.log("Timeout ran at", Date.now() - start);
+  console.log("超时在这时运行", Date.now() - start);
 }, 20);
 while (Date.now() < start + 50) {}
-console.log("Wasted time until", Date.now() - start);
-// → Wasted time until 50     浪费时间至50毫秒之后
-// → Timeout ran at 55        超时在第55毫秒时运行
+console.log("浪费时间至", Date.now() - start);
+// → 浪费时间至50毫秒之后
```
Comment (ghost) [inline 11_async.md]: these lines don't match with the console from 728 & 725 respectively. maybe rephrase it to something along the lines of ”超时运行发生在“? so 730 can be "超时运行发生在 55", and 729 can simply be "浪费时间至50"

Comment (Zheng7426) [reply 11_async.md]: Ok. 👌

**Review by ghost — APPROVED** [review]
